### ISSUE TITLE
Performance / WinError 10054 / Threads

### ISSUE THREAD

**Issue by Postlagerkarte** [genotrance/px#96]
Running px with the default 5 threads and executing an npm command with px set as proxy results in slow execution times. If I increase the number of threads execution time is back to normal.

The log give some hints:

A lot of entries like this->
MainProcess: Thread_13: 1568719470: handle_one_request: Socket error: [WinError 10054] An existing connection was forcibly closed by the remote host
[....]
And a lot of entries like this ->
MainProcess: Thread_19: 1568719456: do_CONNECT: Proxy connection timeout

Any idea if this is related to the configuration of the cooperate proxy or any other idea why this is happening? 

I assume that running px with --threads=500 is not the best idea :)

**Comment by msmicker** [comment]
I am perhaps having a similar issue. px-proxy seems to work well with all of the tools I am using except for npm. With cntlm a test command like 'npm install webpack' will finish relatively quickly (<1min). With px-proxy this never seems to finish. I have debug on and there is a lot of logging but even after ~15 minutes it is not even starting to create files in a node_modules subdirectory.

**Comment by Postlagerkarte** [comment]
@msmicker : if you run px with --threads=500 does it work for you?

**Comment by msmicker** [comment]
It does not seem to, or at least it is still much slower than cntlm.

I tried yarn which works with px-proxy (same config) and this does complete however there are many stacktrace errors throughout. I also noticed px-proxy stacktrace errors in logs with other tools like python pip. The odd thing is that yarn and pip both complete a test package install and with reasonable performance.

The stacktraces run together, but here is an example:
AttributeError: 'list' object has no attribute 'clear'
  File "c:\python27\Lib\SocketServer.py", line 334, in finish_request
  File "c:\users\e0426558\home\px\.env\lib\site-packages\px.py", line 778, in do_CONNECT
----------------------------------------    self.RequestHandlerClass(request, client_address, self)
    rlist.clear()

AttributeError: 'list' object has no attribute 'clear'
  File "c:\python27\Lib\SocketServer.py", line 655, in __init__
----------------------------------------    self.handle()

  File "c:\python27\Lib\BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "c:\users\e0426558\home\px\.env\lib\site-packages\px.py", line 356, in handle_one_request
    httpserver.SimpleHTTPRequestHandler.handle_one_request(self)
  File "c:\python27\Lib\BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "c:\users\e0426558\home\px\.env\lib\site-packages\px.py", line 778, in do_CONNECT
    rlist.clear()
AttributeError: 'list' object has no attribute 'clear'

**Comment by msmicker** [comment]
Brief update. 500 threads did let it complete in 5m52s. The exact same package request by cntlm was 2m 16s. I really like the features of px-proxy and hope I can tune the performance and understand the stacktraces appearing in the logs (usually appearing with pip or yarn).

**Comment by bdwyertech** [comment]
late to the party, but I wrote something similar to PX in Go.  You might want to give this a shot if you need the performance, curious how well/bad it benchmarks for you @msmicker 

https://github.com/bdwyertech/gontlm-proxy

**Comment by genotrance** [comment]
It's been a long time but I just tested npm against px in the default configuration (5 threads, 2 workers) and had no issues. Hoping someone can get back and let me know if you are still seeing this.

I tested npm v8.3.1 on Windows, not sure if that makes any difference though.

**Comment by DubZyy** [comment]
I have a similar issue on a customer environment. Security rules of customers are to not safe password data in a text file. Therefore I'm using px which redirects to customer proxy.
When want to download angular cli via npm, I ran into timeout issues (I guess). The exact error response was a ECONNECTREFUSE after 1-2 minutes.
Watching the system seems to create hundreds of connections, most of them in state "TimeWait"
![image](https://user-images.githubusercontent.com/37286452/166955284-d619cf6f-0811-40e1-bd18-8091e2352d72.png)
peak open connection were 900. average is about 500. I'm not sure if px is the problem or the actual customer proxy.

I had to modify npm config parameters like fetch-retries, timeouts etc. Since then, finally I was at least able to download without npm is running into an error - but it took 20mins (compared my own VM just 2mins)
